File: 26-Nov-2024/Sliding Window Maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        deque<int> Q;
        vector<int> res;
        for (int i = 0; i < k; i++) {
            while (size(Q) > 0 && nums[i] > Q.back()) {
                Q.pop_back();
            }

            Q.push_back(nums[i]);
        }

        res.push_back(Q.front());
        int l = 0;
        for (int r = k; r < size(nums); r++) {
            if (nums[l++] == Q.front()) Q.pop_front();

            cout << endl;
            while (!Q.empty() && nums[r] > Q.back()) {
                Q.pop_back();
            }

            Q.push_back(nums[r]);
            res.push_back(Q.front());
        }

        return res;
    }
};

File: 26-Nov-2024/Sliding Window Maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        counter = defaultdict(int)
        heap = []
        for i in range(k):
            counter[nums[i]] += 1
            heappush(heap, -nums[i])
        
        max_window = [-heap[0]]
        l = 0
        for r in range(k, len(nums)):
            counter[nums[l]] -= 1
            heappush(heap, -nums[r])
            counter[nums[r]] += 1
            while heap and counter[-heap[0]] == 0:
                heappop(heap)
            
            max_window.append(-heap[0])
            l += 1
        
        return max_window